src/Storages/ExternalStream/Kafka/KafkaSource.cpp

```diff
@@ -272,10 +276,17 @@ void KafkaSource::initFormatExecutor(const Kafka * kafka)
     if (!data_format.empty())
     {
         auto input_format
-            = FormatFactory::instance().getInputFormat(data_format, read_buffer, physical_header, query_context, max_block_size);
+            = FormatFactory::instance().getInputFormat(data_format, read_buffer, non_virtual_header, query_context, max_block_size);
 
         format_executor = std::make_unique<StreamingFormatExecutor>(
-            physical_header, std::move(input_format), [](const MutableColumns &, Exception &) -> size_t { return 0; });
+            non_virtual_header, std::move(input_format), [](const MutableColumns &, Exception &) -> size_t { return 0; });
+
+        auto converting_dag = ActionsDAG::makeConvertingActions(
+            non_virtual_header.cloneEmpty().getColumnsWithTypeAndName(),
+            physical_header.cloneEmpty().getColumnsWithTypeAndName(),
+            ActionsDAG::MatchColumnsMode::Name);
+
+        convert_non_virtual_to_physical_action = std::make_shared<ExpressionActions>(std::move(converting_dag));
```

Comment:
This logic has nothing to do with the format executor, extract them into another method, or simply put them in ctor?

Reply:
This is binding with format. In this version, formatter will parse message to the full columns set which `create` statement specified. And this dag action will choose a subset of columns which is needed by the following processors. Thus, when we use a format like csv, it can parse the data as what we want to do.
If it's only a raw data column, we don't need to init this dag action.